tests/rgw/test_s3.py

```diff
@@ -18,101 +44,247 @@ def run(**kw):
+    secure = config.get("is_secure", "no")
+    port = "443" if secure.lower() == "yes" else rgw_frontend_port(nodes, build)
+    create_s3_conf(nodes, build, host, port, secure)
+
+    if not build.startswith("5"):
+        open_firewall_port(rgw_node, port=port, protocol="tcp")
+
+    add_lc_debug(nodes, build)
+
+
+def execute_s3_tests(node: CephNode, build: str) -> int:
     """
-    Performs initial setup and configuration for s3 tests on client node.
+    Return the result of S3 test run.
 
     Args:
-        client_node: The node configured with 'client'
-        rgw_node: The node configured with 'radosgw'
-        config: test configuration
-        build: rhcs4 build version
+        node: The node from which the test execution is triggered.
+        build: the RH build version
 
     Returns:
-        None
+        0 - Success
+        1 - Failure
+    """
+    log.info("Executing s3-tests")
+    try:
+        if build.startswith("4"):
+            cmd = "cd s3-tests; S3TEST_CONF=config.yaml ~/virtualenv/bin/nosetests -v "
+            extra_args = "-a '!fails_on_rgw,!fails_strict_rfc2616,!encryption' s3tests"
+            cmd += extra_args
+        elif build.startswith("5"):
+            cmd = "cd s3-tests; S3TEST_CONF=config.yaml virtualenv/bin/nosetests -v"
+            tags = "-a '!fails_on_rgw,!fails_strict_rfc2616,!encryption'"
+            tests = "s3tests_boto3"
+            cmd = f"{cmd} {tags} {tests}"
+        else:
+            cmd = "cd s3-tests; S3TEST_CONF=config.yaml ~/virtualenv/bin/nosetests -v"
+            extra_args = ["-a '!fails_on_rgw,!lifecycle'"]
+            cmd += " ".join(extra_args)
+
+        out, err = node.exec_command(cmd=cmd, timeout=3600)
+        log.info(out.read().decode())
+        log.info(err.read().decode())
+        return 0
+    except CommandFailed as e:
+        log.warning("Received CommandFailed")
+        log.warning(e)
+        return 1
 
+
+def execute_teardown(nodes: List[CephNode], build: str) -> None:
+    """
+    Execute the test teardown phase.
+    """
+    node = get_node_by_role(nodes, role="client")
+    command = "rm -r s3-tests virtualenv"
+
+    if build.startswith("5"):
+        command = "rm -r s3-tests"
+
+    node.exec_command(cmd=command)
+    del_lc_debug(nodes, build)
+
+
+# Internal functions
+
+
+def get_node_by_role(nodes: List[CephNode], role: str) -> CephNode:
+    """
+    Return the Node that matches the given role.
+
+    Args:
+        nodes:  List of the nodes participating in the cluster
+        role:   The role of the node to be retrieved.
     """
-    log.info("Removing existing s3-tests directory if it exists")
-    client_node.exec_command(cmd="if test -d s3-tests; then rm -r s3-tests; fi")
+    for node in nodes:
+        if node.role == role:
+            return node
+
 
-    log.info("Cloning s3-tests repository")
-    branch = config.get("branch", "ceph-luminous")
+def clone_s3_tests(node: CephNode, branch="master") -> None:
```

Comment:
we already take the branch value from the `config` do we still want to add the default value as `master` here again

Reply:
As I noticed that the branch was not created, I wanted a mechanism to pick the latest... hence the default value. This allows us to avoid specifying the branch name incase of dev runs...